Add object tracking functionality for Python 3.13.3+

pablogsal opened the pull request:
This commit adds support for tracking Python objects that survive test execution
using the new reference tracking API introduced in memray (commit e567c3cf).

*Issue number of the reported bug or feature request: #<number>*

**Describe your changes**
A clear and concise description of the changes you have made.

**Testing performed**
Describe the testing you have performed to ensure that the bug has been addressed, or that the new feature works as planned.

**Additional context**
Add any other context about your contribution here.

pablogsal committed:
Add object tracking feature for Python 3.13.3+

This commit adds support for tracking Python objects that survive test
execution using the new reference tracking API introduced in Memray
(commit e567c3cf).

godlygeek commented on src/pytest_memray/marks.py:
```diff
@@ -267,9 +268,142 @@ def limit_leaks(
+            f"{padding}Object types that leaked (total: {len(self.surviving_objects)} objects):"
+        )
+        for obj_type, count in type_counts.most_common(10):
+            text_lines.append(f"{padding*2}- {obj_type}: {count} instance(s)")
```
It might be better to do:
```suggestion
            text_lines.append(f"{padding*2}- {count} {obj_type} instance(s)")
```
That will make it easier to compare the quantities across different lines. Compare:
```text
- int: 98175 instance(s)
- tuple: 12101 instance(s)
- dict: 8078 instance(s)
```
versus:
```text
- 98175 int instance(s)
- 12101 tuple instance(s)
- 8078 dict instance(s)
```
It's a lot easier to compare the numbers when they're aligned. Maybe the counts should even be right aligned?

godlygeek commented on src/pytest_memray/marks.py:
```diff
@@ -267,9 +267,142 @@ def limit_leaks(
+    """Track objects that survive the test execution."""
+    if _surviving_objects is None:
+        return None
+
```
```suggestion
```
the below `if` already handles everything falsey, so no need to have this check separately.

godlygeek commented on tests/test_pytest_memray.py:
```diff
@@ -271,11 +271,16 @@ def test_foo():
     )
 
     output = result.stdout.str()
-    mock.assert_called_once_with(
```
Why did this need to change? Is this a holdover from an earlier iteration, perhaps? It seems to me like this old assertion should work fine with the currently proposed implementation.

godlygeek commented on tests/test_object_tracking.py:
```diff
@@ -0,0 +1,226 @@
+        output = result.stdout.str()
+
+        assert "CustomClass" in output or "test_various_types" in output
+        assert "set:" in output or "dict:" in output
```
??? Why `or` on both of these lines?

godlygeek commented on tests/test_object_tracking.py:
```diff
@@ -0,0 +1,226 @@
+
+        # Should show summary
+        assert "dict:" in output
+        assert "100" in output or "instance(s)" in output
```
Why `or` again?

godlygeek commented on tests/test_object_tracking.py:
```diff
@@ -0,0 +1,226 @@
+                test_get_leaks._kept = [leaked_list, leaked_dict]
+
+                # Note: The actual implementation would need to inject
+                # a function to retrieve the objects
```
I don't understand how this is meant to work at all 😓

Inject how?

godlygeek commented on src/pytest_memray/marks.py:
```diff
@@ -271,7 +272,7 @@ def limit_leaks(
     """Type that holds information about objects that survived tracking."""
 
-    surviving_objects: list
+    surviving_objects: list[Any]
```
Shouldn't it be `list[object]`, not `list[Any]`? Likewise for every occurrence of `Any` that you've added.

godlygeek commented on src/pytest_memray/plugin.py:
```diff
@@ -213,7 +213,7 @@ def memory_reporting() -> Generator[None, None, None]:
                 tracker_kwargs["track_object_lifetimes"] = True
 
-            tracker = Tracker(result_file, **tracker_kwargs)
+            tracker = Tracker(result_file, **tracker_kwargs)  # type: ignore[call-overload]
```
Huh, what error is it complaining about here?

godlygeek commented:
@pablogsal ping (note that this will need to be rebased in order for tests to pass)

pablogsal commented:
@godlygeek I think I addressed everything. Ready for another round

godlygeek commented:
@pablogsal OK, the ball is back in your court. Take a look at my fixup commits and see if there's anything you're not happy with.

I figured out what was going on with the frame showing up as leaked - we were seeing our own call to `contextlib.contextmanager.__exit__` for when we resumed our generator to stop the tracking. I've worked around that by pushing the tracking disablement up one layer, into our wrapper functions. It's a bit ugly, since it adds some duplication, but it fixes the problem.

I've also added a bit of documentation, and cleaned a few small things up.

I think we probably need to add a `filter_fn` argument to this marker, like the one that `limit_leaks` takes, so that people can ignore known leaks due to caches or the like. Do you agree?

godlygeek committed:
Allow filtering leaked object reports

Since it's very difficult to guarantee a Python function won't leak any
objects, give test authors the ability to suppress leak reports for
objects of their choice.

Signed-off-by: Matt Wozniski <godlygeek@gmail.com>

godlygeek commented:
I've pushed another commit that adds filtering.

One last thing I'm not totally happy with here: I don't like the marker name. `track_leaked_objects` doesn't make it obvious the test will fail if there are any leaked objects. Maybe we should call it `limit_leaked_objects` for consistency with `limit_memory` and `limit_leaks`? Or `ensure_no_leaked_objects` maybe...

pablogsal commented:
Hey @godlygeek ,

I pushed 8 fixups that autosquash into your two commits. Nothing structural, mostly tightening things up, but there were a couple of real bugs worth calling out.

The big one: the Manager was holding onto every "leaked" object for the entire run. It did a `.get` on the surviving-objects dict and never cleared the entry, so the leak detector was itself leaking, and worse, those objects (and everything they referenced) stayed alive through all the later tests, which could quietly skew their results. Changed it to `.pop` so the objects get released once the report's built.

Second real one: the Python 3.13.3+ guard was raising inside the `pytest_pyfunc_call` hookwrapper. That fires during the call phase and raises before the yield, which breaks the wrapper protocol, so on an old interpreter you'd get an ugly internal error mid-run instead of the clean failure the docs promise. Moved it into `pytest_collection_modifyitems` raising `UsageError`, so now it fails at collection like it's supposed to.

The rest is just small stuff. Smoke-tested every path on 3.14.5 with memray 1.19.2, basic leak/no-leak, the filter, the report overflow, repr failures, async, and the version guard. All good shit.

godlygeek commented:
All of your changes LGTM. Thoughts on the marker name? `@pytest.mark.track_leaked_objects` doesn't imply anything about the test failing if there are leaks. I propose changing it to `limit_leaked_objects` - any objections?

pablogsal commented:
No, I think that's a good name will make the change and land

godlygeek reviewed (dismissed):
Approved, pending that rename
